tareas/Tarea1_Andres/intervalos.jl

```diff
@@ -0,0 +1,341 @@
+### A Pluto.jl notebook ###
+# v0.17.1
+
+module Intervalos
+
+export Intervalo, intervalo_vacio
+export isinterior, ⪽, hull, ⊔
+
+# ╔═╡ 6a13438f-43df-42ac-861f-e3522b35b9a1
+struct Intervalo{T<:Real}
+	
+	infimo::T;   supremo::T
+	
+# To promote every subtype of Real to AbstractFloat, and then promote them to the one with more hierarchy, like Float64 or BigFloat
+	function Intervalo(infimo::T1, supremo::T2) where {T1<:Real, T2<:Real}
+		inf, sup, _ = promote(infimo, supremo, 1.0)
+		return Intervalo(inf, sup)
+	end
+
+# To check that the interval satisfy the condition a ≤ b for Intervalo(a,b) ≡ [a,b]
+	function Intervalo(infimo::T, supremo::T) where {T<:AbstractFloat}
+		@assert infimo ≤ supremo
+		return new{T}(infimo, supremo)
+	end
+
+# To allow that Intervalo(a) ≡ [a] → [a, a] also work
+	function Intervalo(same::T) where {T<:Real}
+		return Intervalo(same, same)
+	end
+end
+
+# ╔═╡ d5dd2e80-03a2-41f3-9919-6f88bf385887
+begin   ### [a,b] == [c,d] ⇒ a = b && b = d
+	import Base: ==
+	function ==(I1::Intervalo, I2::Intervalo)
+		inf1, inf2 = promote(I1.infimo, I2.infimo)
+		sup1, sup2 = promote(I1.supremo, I2.supremo)
+		return (inf1 == inf2) && (sup1 == sup2)
+	end
+end
+
+# ╔═╡ aef78ea7-00db-4f69-a67a-3090b1f72e26
+begin   ### Benet's empty interval definition. At the end, I think it works better
+	function intervalo_vacio(arg)
+		if arg == BigFloat
+			return Intervalo(big(Inf), big(Inf))
+		else
+			return Intervalo(Inf, Inf)
+		end
+	end
+	function intervalo_vacio()
+		return Intervalo(Inf, Inf)
+	end
```

Comment:
Esto no lo noté antes, pero este tipo de `if`s que checan el tipo no son correctos en Julia, y para eso tenemos lo que es *dispatch*. (El problema es que tu función deja de ser *type stable*.) Esta función la puedes reescribir como:
```julia
function intervalo_vacio(arg::T) where {T<:Real}
    TT = promote_type(arg, Float64) # haz la promoción al menos a Float64
    return Intervalo(T(Inf), T(Inf))
end
```
lo que es más conciso y fácil de leer.